sw/device/lib/dif/dif_kmac_unittest.cc

```diff
@@ -717,4 +738,213 @@ TEST_F(KmacStatusTest, BadArg) {
+
+  EXPECT_DIF_BADARG(dif_kmac_get_error(&kmac_, nullptr));
+}
+
+class KmacSqueezeTest : public KmacTest {
+ protected:
+  uint32_t out_buffer_[8];
+  size_t processed_;
+  static constexpr std::array<std::array<uint32_t, 64>, 2> kOutShares = {
+      0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0x5AA5A55A, 0x5AA5A55A,
+      0x5AA5A55A, 0x5AA5A55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A,
+      0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A, 0xA55AA55A, 0xA55AA55A,
+      0xA55AA55A, 0xA55AA55A, 0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A,
+      0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0x5AA5A55A, 0x5AA5A55A,
+      0x5AA5A55A, 0x5AA5A55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A,
+      0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A, 0xA55AA55A, 0xA55AA55A,
+      0xA55AA55A, 0xA55AA55A, 0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A,
+      0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0x5AA5A55A, 0x5AA5A55A,
+      0x5AA5A55A, 0x5AA5A55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A, 0xA55AA55A,
+      0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A, 0x5AA5A55A};
+
+  KmacSqueezeTest() {
+    op_state_.r = GetRateWords(256);
+    op_state_.squeezing = false;
+    op_state_.append_d = false;
+  }
+
+  void ExpectAppendSize() {
+    uint32_t d = op_state_.d * sizeof(uint32_t) * 8;
+    uint8_t len = 1 + (d > 0xFF) + (d > 0xFFFF) + (d > 0xFFFFFF);
+    uint8_t shift = len * 8;
+    do {
+      shift -= 8;
+      EXPECT_WRITE8(KMAC_MSG_FIFO_REG_OFFSET, (d >> shift) & 0xFF);
+    } while (shift);
+    EXPECT_WRITE8(KMAC_MSG_FIFO_REG_OFFSET, len & 0xFF);
+  }
+
+  void ExpectReadOutput(const uint32_t *share1, const uint32_t *share2,
+                        size_t len) {
+    uint32_t offset = KMAC_STATE_REG_OFFSET;
+    for (size_t i = 0; i < len; ++i) {
+      // Read both shares from state register and combine using XOR.
+      EXPECT_READ32(offset, share1[i]);
+      EXPECT_READ32(offset + 0x100, share2[i]);
+      offset += sizeof(uint32_t);
+    }
+  }
+};
+constexpr std::array<std::array<uint32_t, 64>, 2> KmacSqueezeTest::kOutShares;
+
+TEST_F(KmacSqueezeTest, GenerateExtraStatesSuccess) {
+  uint32_t out_buffer[64];
+  op_state_.d = ARRAYSIZE(out_buffer);
+
+  EXPECT_WRITE32(KMAC_CMD_REG_OFFSET,
+                 {{KMAC_CMD_CMD_OFFSET, KMAC_CMD_CMD_VALUE_PROCESS}});
+  EXPECT_READ32(KMAC_STATUS_REG_OFFSET, {{KMAC_STATUS_SHA3_SQUEEZE_BIT, true}});
+  ExpectReadOutput(kOutShares[0].data(), kOutShares[1].data(), 34);
+  EXPECT_WRITE32(KMAC_CMD_REG_OFFSET,
+                 {{KMAC_CMD_CMD_OFFSET, KMAC_CMD_CMD_VALUE_RUN}});
+  EXPECT_READ32(KMAC_STATUS_REG_OFFSET, {{KMAC_STATUS_SHA3_SQUEEZE_BIT, true}});
+  ExpectReadOutput(&kOutShares[0].data()[34], &kOutShares[1].data()[34],
+                   kOutShares[0].size() - 34);
+
+  EXPECT_DIF_OK(dif_kmac_squeeze(&kmac_, &op_state_, out_buffer,
+                                 ARRAYSIZE(out_buffer), nullptr));
+
+  EXPECT_EQ(op_state_.squeezing, true);
```

Comment:
Maybe it's worth defining an `operator==` and `operator<<` for this type at the top of the file (in global scope, so ADL can find it). That way you can done one EXPECT_EQ. I'd take a look at the i2c tests for this.

Reply:
Neat. Done